File: learn-react/TweetApp/backend/src/routes/topic/Topic.service.js
// Topic.service.js

const { Topic, TopicSection } = require("./Topic.model");
const { v4: uuidv4 } = require("uuid");

const createTopic = async (topicData) => {
  return await Topic.create(topicData);
};

/**
 * Create multiple topics in one request.
 * @param {Array<{ name: string, parentId?: string, description?: string, smartContent?: object, tags?: string[], occurenceDate?: string|Date }>} topicPayloads
 * @returns {{ created: Array, errors: Array<{ index: number, message: string }> }}
 */
const createTopicsBulk = async (topicPayloads) => {
  if (!Array.isArray(topicPayloads) || topicPayloads.length === 0) {
    return { created: [], errors: [] };
  }
  const created = [];
  const errors = [];
  for (let i = 0; i < topicPayloads.length; i++) {
    const payload = topicPayloads[i];
    try {
      const doc = {
        name: payload.name != null ? String(payload.name).trim() : "",
        parentId: payload.parentId != null ? payload.parentId : "",
        description: payload.description != null ? payload.description : undefined,
        smartContent: payload.smartContent != null ? payload.smartContent : undefined,
        tags: Array.isArray(payload.tags) ? payload.tags : [],
        occurenceDate:
          payload.occurenceDate != null
            ? new Date(payload.occurenceDate)
            : new Date(),
      };
      if (!doc.name) {
        errors.push({ index: i, message: "Name is required" });
        continue;
      }
      const topic = await Topic.create(doc);
      created.push(topic.toObject ? topic.toObject() : topic);
    } catch (err) {
      errors.push({
        index: i,
        message: err.message || "Failed to create topic",
      });
    }
  }
  return { created, errors };
};

const updateTopicByUniqueId = async (uniqueId, topicData) => {
  try {
    let topic = await Topic.findOne({ uniqueId });
    if (!topic) {
      throw new Error("Task not found");
    }
    // console.log(`topicData : ${JSON.stringify(topicData)}`)
    const {
      parentId,
      name,
      description,
      smartContent,
      tags,
      occurenceDate,
      children,
    } = topicData;
    if (topicData.published === true) {
      throw new Error("Use PUT /:uniqueId/publish to publish a topic");
    }
    if (topicData.published === false) {
      topic.published = false;
    }
    topic.parentId = parentId || topic.parentId;
    topic.name = name || topic.name;
    topic.description = description || topic.description;
    topic.tags = tags && tags.length >= 0 ? tags : topic.tags;
    topic.occurenceDate =
      occurenceDate != null ? occurenceDate : topic.occurenceDate;
    topic.smartContent = smartContent || topic.smartContent;
    topic.updatedDate = new Date();
    topic = await topic.save();
    const childrenIds = children || [];
    // console.log(`childrenIds: ${childrenIds}`)
    await Topic.updateMany(
      { uniqueId: { $in: childrenIds } },
      { parentId: topic.uniqueId }
    );
    return topic;
  } catch (err) {
    console.error(err);
    throw err;
  }
};

const getAllTopics = async () => {
  try {
    const selectFields = {
      uniqueId: 1,
      name: 1,
      parentId: 1,
      description: 1,
      smartContent: 1,
      tags: 1,
      published: 1,
    };
    // console.log(`[Topic.service]: [getAllTopics]: Going to fetch all topics`);
    const topics = await getTopics(null, { ...selectFields });
    return topics;
  } catch (err) {
    console.error(err);
    throw err;
  }
};

/**
 * Flatten tree of topics into a single array (for export).
 * Each item has uniqueId, name, parentId, description, smartContent, tags, ancestors, sections.
 */
function flattenTopics(tree, ancestors = []) {
  if (!tree || !Array.isArray(tree)) return [];
  const list = [];
  for (const node of tree) {
    const item = {
      uniqueId: node.uniqueId,
      name: node.name,
      parentId: node.parentId || null,
      description: node.description != null ? node.description : "",
      smartContent: node.smartContent != null ? node.smartContent : null,
      tags: node.tags || [],
      published: node.published === true,
      ancestors: ancestors.map((a) => ({ uniqueId: a.uniqueId, name: a.name })),
      sections: node.sections || [],
    };
    list.push(item);
    if (node.children && node.children.length > 0) {
      const nextAncestors = [...ancestors, { uniqueId: node.uniqueId, name: node.name }];
      list.push(...flattenTopics(node.children, nextAncestors));
    }
  }
  return list;
}

const getAllTopicsFlat = async () => {
  const tree = await getAllTopics();
  return flattenTopics(tree);
};

const sectionExportFields = {
  uniqueId: 1,
  linkedTopicUniqueId: 1,
  name: 1,
  smartContent: 1,
  order: 1,
  softDelete: 1,
  tags: 1,
  createdDate: 1,
  updatedDate: 1,
};

/**
 * Attach topic sections to each node in the tree (for export). Mutates nodes.
 */
async function enrichTreeWithSections(tree) {
  if (!tree || !Array.isArray(tree)) return tree;
  await Promise.all(
    tree.map(async (node) => {
      const sections = await TopicSection.find({
        linkedTopicUniqueId: node.uniqueId,
      })
        .select(sectionExportFields)
        .lean();
      node.sections = sections || [];
      if (node.children && node.children.length > 0) {
        await enrichTreeWithSections(node.children);
      }
    })
  );
  return tree;
}

/**
 * Topics tree with sections attached (for export).
 */
const getAllTopicsForExport = async () => {
  const tree = await getAllTopics();
  return enrichTreeWithSections(tree);
};

/**
 * Flat list of topics with sections on each topic (for export).
 */
const getAllTopicsFlatForExport = async () => {
  const tree = await getAllTopicsForExport();
  return flattenTopics(tree);
};

async function getTopics(parentId, selectFields, options = {}) {
  try {
    const criteria = parentId
      ? { parentId }
      : { parentId: { $in: [null, undefined, ""] } };
    if (options.publishedOnly) {
      criteria.published = true;
    }
    let topics = await Topic.find(criteria).select(selectFields);
    const topicsWithChildren = await Promise.all(
      topics.map(async (topic) => {
        const children = await getTopics(topic.uniqueId, selectFields, options);
        const ancestors = await getAllAncestors(topic.parentId);
        return { ...topic.toObject(), children, ancestors };
        // return { ...topic.toObject(), children };
      })
    );
    return topicsWithChildren;
  } catch (error) {
    console.error(error);
    return [];
  }
}

const getPublishedTopics = async () => {
  try {
    const selectFields = {
      uniqueId: 1,
      name: 1,
      parentId: 1,
      description: 1,
      smartContent: 1,
      tags: 1,
      published: 1,
    };
    const flatList = await Topic.find({ published: true })
      .select(selectFields)
      .lean();
    return buildTreeFromFlatPublished(flatList);
  } catch (err) {
    console.error(err);
    throw err;
  }
};

/**
 * Build a tree from a flat list of published topics.
 * Topics whose parent is not in the list (or has no parent) appear as roots.
 */
function buildTreeFromFlatPublished(flatList) {
  if (!flatList || flatList.length === 0) return [];
  const byId = new Map(flatList.map((t) => [t.uniqueId, { ...t, children: [] }]));
  const roots = [];
  for (const t of flatList) {
    const node = byId.get(t.uniqueId);
    const parentId = t.parentId || null;
    const parent = parentId ? byId.get(parentId) : null;
    if (!parent) {
      roots.push(node);
    } else {
      parent.children.push(node);
    }
  }
  return roots;
}

const getTopicByUniqueId = async (uniqueId) => {
  try {
    const selectFields = {
      uniqueId: 1,
      name: 1,
      parentId: 1,
      tags: 1,
    };

    const topic = await Topic.findOne({ uniqueId });

    if (!topic) {
      throw new Error("Topic not found , uniqueId : " + uniqueId);
    }

    const sections = await TopicSection.find({
      linkedTopicUniqueId: topic.uniqueId,
    }).select({
      uniqueId: 1,
      name: 1,
      linkedTopicUniqueId: 1,
      order: 1,
    });

    // const children = await Topic.find({ parentId: uniqueId });
    const children = await getTopics(topic.uniqueId, selectFields);
    let ancestors = [];
    try {
      ancestors = await getAllAncestors(topic.parentId);
      // res.json(ancestors);
    } catch (error) {
      // res.status(500).json({ error: error.message });
      console.log(error);
      ancestors = [];
    }

    const responseDTO = {
      ...topic.toObject(),
      children: children.map((child) => ({
        name: child.name,
        uniqueId: child.uniqueId,
        children: child.children,
      })),
      sections: sections
        ? sections.map((s) => ({
          uniqueId: s.uniqueId,
          name: s.name,
          linkedTopicUniqueId: s.linkedTopicUniqueId,
          order: s.order,
        }))
        : [],
      ancestors: [...ancestors],
    };
    return responseDTO;
  } catch (err) {
    console.error(err);
    throw err;
  }
};

// Recursive function to get all ancestors of a link
async function getAllAncestors(parentId, ancestors = []) {
  // console.log(`start: parentId: ${parentId} :   function getAllAncestors : ${JSON.stringify(ancestors)}`);
  if (!parentId) {
    return ancestors;
  }
  const topic = await Topic.findOne({ uniqueId: parentId });
  // console.log(`topic: ${JSON.stringify(topic)}`);
  // if (!topic || !topic.parentId) {
  //   return ancestors;
  // }
  ancestors.unshift({
    name: topic.name,
    uniqueId: topic.uniqueId,
    parentId: topic.parentId,
  }); // Add the name of the current topic to ancestors array
  // console.log(`final: function getAllAncestors : ${JSON.stringify(ancestors)}`);
  return getAllAncestors(topic.parentId, ancestors); // Recursively call to get ancestors of the parent topic
}

/**
 * Returns true only if the topic with given parentId and all its ancestors are published.
 * Root (no parent) is considered "published" for the purpose of allowing root topics to be published.
 */
async function areAllAncestorsPublished(parentId) {
  if (!parentId) return true;
  const parent = await Topic.findOne({ uniqueId: parentId }).select({ published: 1, parentId: 1 });
  if (!parent) return true;
  if (!parent.published) return false;
  return areAllAncestorsPublished(parent.parentId);
}

/**
 * Publish a topic by uniqueId. Succeeds only if the topic's parent and all ancestors are already published.
 */
async function publishTopicByUniqueId(uniqueId) {
  const topic = await Topic.findOne({ uniqueId });
  if (!topic) {
    throw new Error("Topic not found, uniqueId: " + uniqueId);
  }
  if (topic.published) {
    return topic;
  }
  const canPublish = await areAllAncestorsPublished(topic.parentId);
  if (!canPublish) {
    throw new Error("Cannot publish: parent or an ancestor topic is not published");
  }
  topic.published = true;
  topic.updatedDate = new Date();
  await topic.save();
  return topic;
}

const searchTopics = async (searchString, searchOptions) => {
  const regex = new RegExp(searchString, "i"); // 'i' for case insensitive

  const selectFields = {
    uniqueId: 1,
    name: 1,
    parentId: 1,
    tags: 1,
  };

  const criteria = {
    $or: [
      { name: { $regex: regex } },
      //{ description: { $regex: regex } },
    ],
  };

  if (
    searchOptions &&
    searchOptions.description &&
    searchOptions.description > 0
  ) {
    selectFields.description = 1;
    criteria["$or"].push({ description: { $regex: regex } });
  }

  return await Topic.find(criteria).select(selectFields);
};

const createTopicSection = async (sectionData) => {
  // console.log(`[Topic.service]: sectionData : ${JSON.stringify(sectionData)}`);
  const { linkedTopicUniqueId, name, smartContent, order, tags } = sectionData;

  const uniqueId = uuidv4();
  const newTopicSection = {
    uniqueId,
    linkedTopicUniqueId,
    name,
    smartContent,
    order: order && typeof order === "number" ? order : 9999,
    tags: tags && tags.length > 0 ? tags : [],
  };
  await TopicSection.create(newTopicSection);
  return newTopicSection;
};

const getAllTopicSectionsById = async (linkedTopicUniqueId) => {
  const sections = await TopicSection.find({
    linkedTopicUniqueId: linkedTopicUniqueId,
  });
  return sections;
};

const getTopicSectionsById = async (linkedTopicUniqueId, sectionUniqueId) => {
  const section = await TopicSection.findOne({
    linkedTopicUniqueId: linkedTopicUniqueId,
    uniqueId: sectionUniqueId,
  });
  if (!section) {
    throw new Error(
      `TopicSection not found , linkedTopicUniqueId : ${linkedTopicUniqueId}, sectionUniqueId: ${sectionUniqueId}`
    );
  }
  return section;
};

const updateTopicSectionsById = async (
  topicUniqueId,
  sectionUniqueId,
  sectionData
) => {
  console.log(
    `[Topic.service]: [updateTopicSectionsById]: sectionData : ${JSON.stringify(
      sectionData
    )}`
  );
  const { linkedTopicUniqueId, name, smartContent, order, tags } = sectionData;
  try {
    let section = await TopicSection.findOne({
      linkedTopicUniqueId: linkedTopicUniqueId,
      uniqueId: sectionUniqueId,
    });
    if (!section) {
      throw new Error(
        `TopicSection not found , linkedTopicUniqueId : ${linkedTopicUniqueId}, sectionUniqueId: ${sectionUniqueId}`
      );
    }
    // console.log(`sectionData : ${JSON.stringify(sectionData)}`)
    // const { parentId, name, description, tags, occurenceDate, children } = topicData;
    section.smartContent = smartContent || section.smartContent;
    section.name = name || section.name;
    section.order = order || section.order;
    // section.description = description || section.description;
    section.tags = tags && tags.length >= 0 ? tags : section.tags;
    // section.occurenceDate = occurenceDate != null ? occurenceDate : topic.occurenceDate;
    section.updatedDate = new Date();
    section = await section.save();
    return section;
  } catch (err) {
    console.error(err);
    throw err;
  }
};

const getTopicsByTagId = async (tagId) => {
  try {
    const selectFields = {
      uniqueId: 1,
      name: 1,
      //parentId: 1,
      //tags: 1,
    };
    const topics = await Topic.find({ tags: tagId }).select(selectFields);
    return topics;
  } catch (error) {
    console.error(error);    
    throw new Error(`Error retrieving topics with tagId ${tagId}: ${error.message}`);
  }
};

const getTopicSectionsByTagId = async (tagId) => {
  try {
    const selectFields = {
      uniqueId: 1,
      name: 1,
      linkedTopicUniqueId: 1,
      //tags: 1,
    };
    const topicSections = await TopicSection.find({ tags: tagId }).select(selectFields);
    return topicSections;
  } catch (error) {
    console.error(error);    
    throw new Error(`Error retrieving topicSections with tagId ${tagId}: ${error.message}`);
  }
};

module.exports = {
  createTopic,
  createTopicsBulk,
  updateTopicByUniqueId,
  publishTopicByUniqueId,
  getAllTopics,
  getPublishedTopics,
  getAllTopicsFlat,
  getAllTopicsForExport,
  getAllTopicsFlatForExport,
  getTopicByUniqueId,
  searchTopics,
  createTopicSection,
  getAllTopicSectionsById,
  getTopicSectionsById,
  updateTopicSectionsById,
  getTopicsByTagId,
  getTopicSectionsByTagId
};

File: learn-react/TweetApp/backend-spring-boot-3/src/main/java/com/tweetapp/service/TopicService.java
package com.tweetapp.service;

import com.tweetapp.model.Topic;
import com.tweetapp.model.TopicSection;
import com.tweetapp.repository.TopicRepository;
import com.tweetapp.repository.TopicSectionRepository;
import com.tweetapp.util.UuidUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class TopicService {
    
    @Autowired
    private TopicRepository topicRepository;
    
    @Autowired
    private TopicSectionRepository topicSectionRepository;
    
    public Topic createTopic(Map<String, Object> topicData) {
        Topic topic = new Topic();
        topic.setUniqueId(UuidUtil.generateUuid());
        topic.setName((String) topicData.get("name"));
        topic.setDescription((String) topicData.get("description"));
        topic.setParentId(topicData.get("parentId") != null ? (String) topicData.get("parentId") : "");
        topic.setTags((List<String>) topicData.get("tags"));
        topic.setPublished(Boolean.FALSE);
        topic.setOccurenceDate(topicData.get("occurenceDate") != null ? 
            LocalDateTime.parse((String) topicData.get("occurenceDate")) : LocalDateTime.now());
        topic.setCreatedDate(LocalDateTime.now());
        topic.setUpdatedDate(LocalDateTime.now());
        return topicRepository.save(topic);
    }
    
    public Topic updateTopicByUniqueId(String uniqueId, Map<String, Object> topicData) {
        Optional<Topic> optionalTopic = topicRepository.findByUniqueId(uniqueId);
        if (optionalTopic.isPresent()) {
            Topic topic = optionalTopic.get();
            if (Boolean.TRUE.equals(topicData.get("published"))) {
                throw new IllegalArgumentException("Use PUT /:uniqueId/publish to publish a topic");
            }
            if (Boolean.FALSE.equals(topicData.get("published"))) {
                topic.setPublished(Boolean.FALSE);
            }
            if (topicData.containsKey("name")) topic.setName((String) topicData.get("name"));
            if (topicData.containsKey("description")) topic.setDescription((String) topicData.get("description"));
            if (topicData.containsKey("parentId")) topic.setParentId((String) topicData.get("parentId"));
            if (topicData.containsKey("tags")) topic.setTags((List<String>) topicData.get("tags"));
            if (topicData.containsKey("occurenceDate")) {
                topic.setOccurenceDate(LocalDateTime.parse((String) topicData.get("occurenceDate")));
            }
            topic.setUpdatedDate(LocalDateTime.now());
            return topicRepository.save(topic);
        }
        return null;
    }

    public Map<String, Object> createTopicsBulk(List<Map<String, Object>> topicsPayload) {
        List<Topic> created = new ArrayList<>();
        List<Map<String, Object>> errors = new ArrayList<>();
        if (topicsPayload == null || topicsPayload.isEmpty()) {
            return Map.of("created", created, "errors", errors);
        }

        for (int i = 0; i < topicsPayload.size(); i++) {
            try {
                Map<String, Object> payload = topicsPayload.get(i);
                String name = payload.get("name") != null ? String.valueOf(payload.get("name")).trim() : "";
                if (name.isEmpty()) {
                    errors.add(Map.of("index", i, "message", "Name is required"));
                    continue;
                }
                created.add(createTopic(payload));
            } catch (Exception ex) {
                errors.add(Map.of("index", i, "message", ex.getMessage() != null ? ex.getMessage() : "Failed to create topic"));
            }
        }

        return Map.of("created", created, "errors", errors);
    }
    
    public List<Topic> getAllTopics() {
        return topicRepository.findAll();
    }

    public List<Map<String, Object>> getAllTopicsForExport() {
        List<Topic> topics = topicRepository.findAll();
        Map<String, List<TopicSection>> sectionsByTopic = topicSectionRepository.findAll().stream()
            .collect(Collectors.groupingBy(TopicSection::getLinkedTopicUniqueId));

        List<Map<String, Object>> export = new ArrayList<>();
        for (Topic topic : topics) {
            Map<String, Object> row = new HashMap<>();
            row.put("uniqueId", topic.getUniqueId());
            row.put("name", topic.getName());
            row.put("parentId", topic.getParentId());
            row.put("description", topic.getDescription());
            row.put("smartContent", topic.getSmartContent());
            row.put("tags", topic.getTags());
            row.put("published", Boolean.TRUE.equals(topic.getPublished()));
            row.put("sections", sectionsByTopic.getOrDefault(topic.getUniqueId(), List.of()));
            export.add(row);
        }
        return export;
    }

    public List<Map<String, Object>> getAllTopicsFlatForExport() {
        List<Topic> topics = topicRepository.findAll();
        Map<String, Topic> topicByUniqueId = topics.stream()
            .collect(Collectors.toMap(Topic::getUniqueId, t -> t, (a, b) -> a));
        Map<String, List<TopicSection>> sectionsByTopic = topicSectionRepository.findAll().stream()
            .collect(Collectors.groupingBy(TopicSection::getLinkedTopicUniqueId));

        List<Map<String, Object>> flattened = new ArrayList<>();
        for (Topic topic : topics) {
            Map<String, Object> row = new HashMap<>();
            row.put("uniqueId", topic.getUniqueId());
            row.put("name", topic.getName());
            row.put("parentId", topic.getParentId());
            row.put("description", topic.getDescription());
            row.put("smartContent", topic.getSmartContent());
            row.put("tags", topic.getTags());
            row.put("published", Boolean.TRUE.equals(topic.getPublished()));
            row.put("ancestors", getAllAncestors(topic.getParentId(), topicByUniqueId));
            row.put("sections", sectionsByTopic.getOrDefault(topic.getUniqueId(), List.of()));
            flattened.add(row);
        }
        return flattened;
    }

    public List<Map<String, Object>> getPublishedTopics() {
        List<Topic> published = topicRepository.findByPublishedTrue();
        return buildTreeFromFlatPublished(published);
    }
    
    public Topic getTopicByUniqueId(String uniqueId) {
        return topicRepository.findByUniqueId(uniqueId).orElse(null);
    }

    public Topic publishTopicByUniqueId(String uniqueId) {
        Topic topic = topicRepository.findByUniqueId(uniqueId)
            .orElseThrow(() -> new IllegalArgumentException("Topic not found, uniqueId: " + uniqueId));
        if (Boolean.TRUE.equals(topic.getPublished())) {
            return topic;
        }
        if (!areAllAncestorsPublished(topic.getParentId())) {
            throw new IllegalArgumentException("Cannot publish: parent or an ancestor topic is not published");
        }
        topic.setPublished(Boolean.TRUE);
        topic.setUpdatedDate(LocalDateTime.now());
        return topicRepository.save(topic);
    }
    
    public List<Topic> searchTopics(String searchString, Map<String, Object> searchOptions) {
        return topicRepository.findByNameContainingIgnoreCase(searchString);
    }
    
    public Topic createTopicSection(Map<String, Object> sectionData) {
        TopicSection section = new TopicSection();
        section.setUniqueId(UuidUtil.generateUuid());
        section.setLinkedTopicUniqueId((String) sectionData.get("linkedTopicUniqueId"));
        section.setName((String) sectionData.get("name"));
        section.setOrder((Integer) sectionData.get("order"));
        section.setTags((List<String>) sectionData.get("tags"));
        section.setCreatedDate(LocalDateTime.now());
        section.setUpdatedDate(LocalDateTime.now());
        topicSectionRepository.save(section);
        
        // Return the updated topic
        return topicRepository.findByUniqueId(section.getLinkedTopicUniqueId()).orElse(null);
    }
    
    public List<TopicSection> getAllTopicSectionsById(String uniqueId) {
        return topicSectionRepository.findByLinkedTopicUniqueId(uniqueId);
    }
    
    public TopicSection getTopicSectionsById(String uniqueId, String sectionUniqueId) {
        return topicSectionRepository.findByUniqueId(sectionUniqueId).orElse(null);
    }
    
    public Topic updateTopicSectionsById(String uniqueId, String sectionUniqueId, Map<String, Object> sectionData) {
        Optional<TopicSection> optionalSection = topicSectionRepository.findByUniqueId(sectionUniqueId);
        if (optionalSection.isPresent()) {
            TopicSection section = optionalSection.get();
            if (sectionData.containsKey("name")) section.setName((String) sectionData.get("name"));
            if (sectionData.containsKey("order")) section.setOrder((Integer) sectionData.get("order"));
            if (sectionData.containsKey("tags")) section.setTags((List<String>) sectionData.get("tags"));
            section.setUpdatedDate(LocalDateTime.now());
            topicSectionRepository.save(section);
            
            // Return the updated topic
            return topicRepository.findByUniqueId(uniqueId).orElse(null);
        }
        return null;
    }

    private boolean areAllAncestorsPublished(String parentId) {
        if (parentId == null || parentId.isBlank()) return true;
        Topic parent = topicRepository.findByUniqueId(parentId).orElse(null);
        if (parent == null) return true;
        if (!Boolean.TRUE.equals(parent.getPublished())) return false;
        return areAllAncestorsPublished(parent.getParentId());
    }

    private List<Map<String, Object>> getAllAncestors(String parentId, Map<String, Topic> topicByUniqueId) {
        List<Map<String, Object>> ancestors = new ArrayList<>();
        String currentParentId = parentId;
        while (currentParentId != null && !currentParentId.isBlank()) {
            Topic parent = topicByUniqueId.get(currentParentId);
            if (parent == null) break;
            ancestors.add(0, Map.of(
                "uniqueId", parent.getUniqueId(),
                "name", parent.getName()
            ));
            currentParentId = parent.getParentId();
        }
        return ancestors;
    }

    private List<Map<String, Object>> buildTreeFromFlatPublished(List<Topic> flatList) {
        if (flatList == null || flatList.isEmpty()) return List.of();

        Map<String, Map<String, Object>> byId = new HashMap<>();
        for (Topic t : flatList) {
            Map<String, Object> node = new HashMap<>();
            node.put("uniqueId", t.getUniqueId());
            node.put("name", t.getName());
            node.put("parentId", t.getParentId());
            node.put("description", t.getDescription());
            node.put("smartContent", t.getSmartContent());
            node.put("tags", t.getTags());
            node.put("published", Boolean.TRUE.equals(t.getPublished()));
            node.put("children", new ArrayList<Map<String, Object>>());
            byId.put(t.getUniqueId(), node);
        }

        Set<String> publishedIds = byId.keySet();
        List<Map<String, Object>> roots = new ArrayList<>();
        for (Topic t : flatList) {
            Map<String, Object> node = byId.get(t.getUniqueId());
            String parentId = t.getParentId();
            if (parentId == null || parentId.isBlank() || !publishedIds.contains(parentId)) {
                roots.add(node);
                continue;
            }
            Map<String, Object> parent = byId.get(parentId);
            if (parent != null) {
                ((List<Map<String, Object>>) parent.get("children")).add(node);
            } else {
                roots.add(node);
            }
        }
        return roots;
    }
}

